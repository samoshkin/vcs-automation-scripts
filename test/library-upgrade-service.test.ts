import LibraryUpgradeService from '../src/library-upgrade-service';
import * as tmp from 'tmp-promise';
import * as fs from 'fs/promises';
import type { PackageJson } from 'type-fest';

let service: LibraryUpgradeService;
let tmpFile: tmp.FileResult;

beforeEach(async () => {
  service = new LibraryUpgradeService();
});

afterEach(async() => {
  if (tmpFile) {
    tmpFile.cleanup();
  }
});

test('upgrades library version', async () => {
  const path = await useFixtureFile('test/fixtures/package.json');

  await service.upgradeLibrary(path, 'shelljs', '1.0.0');

  const packageJson = await readPackageJson(path);
  expect(packageJson).toMatchSnapshot();
  expect(packageJson.devDependencies['shelljs']).toBe('1.0.0');

  expect(await readPackageJsonRaw(path)).toMatchSnapshot();
});

test('can upgrade version in multiple locations', async () => {
  const path = await useFixtureFile('test/fixtures/package_multiple_locations.json');

  await service.upgradeLibrary(path, 'shelljs', '1.0.0', ['dependencies', 'devDependencies', 'peerDependencies']);

  const packageJson = await readPackageJson(path);
  expect(packageJson).toMatchSnapshot();
  expect(packageJson.devDependencies['shelljs']).toBe('1.0.0');
  expect(packageJson.dependencies['shelljs']).toBe('1.0.0');
  expect(packageJson.peerDependencies['shelljs']).toBe('1.0.0');
});

test('fails if package.json file not found', async () => {
  await expect(service.upgradeLibrary('test/fixture/non-existent-file.json', 'bitbucket', '2.10.0')).rejects
    .toThrow('File at \'test/fixture/non-existent-file.json\' is not found');
});

test('fails to downgrade the library version', async () => {
  const path = await useFixtureFile('test/fixtures/package.json');

  await expect(service.upgradeLibrary(path, 'bitbucket', '2.10.0')).rejects
    .toThrow('Current library version \'^2.11.0\' looks to be greater than supplied version \'2.10.0\'');
});

test('fails if library usage is found', async () => {
  const path = await useFixtureFile('test/fixtures/package.json');

  await expect(service.upgradeLibrary(path, 'axios', '2.10.0')).rejects
    .toThrow('Usage of library \'axios\' is not found in package.json');
});

async function useFixtureFile(path: string): Promise<string> {
  tmpFile = await tmp.file();
  await fs.copyFile(path, tmpFile.path);
  return tmpFile.path;
}

async function readPackageJson(path: string): Promise<PackageJson> {
  return JSON.parse(await fs.readFile(path, 'utf-8'));
}

async function readPackageJsonRaw(path: string): Promise<string> {
  return fs.readFile(path, 'utf-8');
}
