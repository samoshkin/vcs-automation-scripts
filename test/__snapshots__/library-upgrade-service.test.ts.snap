// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can upgrade version in multiple locations 1`] = `
Object {
  "author": "Alexey Samoshkin",
  "dependencies": Object {
    "bitbucket": "^2.11.0",
    "semver": "^7.5.4",
    "shelljs": "1.0.0",
    "tmp": "^0.2.1",
    "ts-node": "^10.9.1",
  },
  "description": "VCS automation scripts",
  "devDependencies": Object {
    "@types/jest": "^27.0.0",
    "@types/node": "^18.15.3",
    "@types/semver": "^7.5.5",
    "@types/shelljs": "^0.8.15",
    "@types/tmp": "^0.2.6",
    "@typescript-eslint/eslint-plugin": "4.14.2",
    "@typescript-eslint/parser": "4.14.2",
    "eslint": "7.19.0",
    "jest": "^27.0.6",
    "shelljs": "1.0.0",
    "ts-jest": "^27.0.3",
    "type-fest": "^4.8.1",
    "typescript": "4.9",
  },
  "engines": Object {
    "node": ">=18.0.0",
  },
  "license": "MIT",
  "name": "vcs-automation-scripts",
  "peerDependencies": Object {
    "shelljs": "1.0.0",
  },
  "repository": Object {
    "type": "git",
    "url": "https://github.com/Microsoft/TypeScript-Node-Starter",
  },
  "scripts": Object {
    "build": "tsc",
    "buildw": "tsc -w",
    "lint": "tsc --noEmit && eslint",
    "start": "node -r ts-node/register src/main.ts",
    "test": "jest --forceExit --verbose",
  },
  "version": "0.0.1",
}
`;

exports[`upgrades library version 1`] = `
Object {
  "author": "Alexey Samoshkin",
  "dependencies": Object {
    "bitbucket": "^2.11.0",
    "semver": "^7.5.4",
    "tmp": "^0.2.1",
    "ts-node": "^10.9.1",
  },
  "description": "VCS automation scripts",
  "devDependencies": Object {
    "@types/jest": "^27.0.0",
    "@types/node": "^18.15.3",
    "@types/semver": "^7.5.5",
    "@types/shelljs": "^0.8.15",
    "@types/tmp": "^0.2.6",
    "@typescript-eslint/eslint-plugin": "4.14.2",
    "@typescript-eslint/parser": "4.14.2",
    "eslint": "7.19.0",
    "jest": "^27.0.6",
    "shelljs": "1.0.0",
    "ts-jest": "^27.0.3",
    "type-fest": "^4.8.1",
    "typescript": "4.9",
  },
  "engines": Object {
    "node": ">=18.0.0",
  },
  "license": "MIT",
  "name": "vcs-automation-scripts",
  "repository": Object {
    "type": "git",
    "url": "https://github.com/Microsoft/TypeScript-Node-Starter",
  },
  "scripts": Object {
    "build": "tsc",
    "buildw": "tsc -w",
    "lint": "tsc --noEmit && eslint",
    "start": "node -r ts-node/register src/main.ts",
    "test": "jest --forceExit --verbose",
  },
  "version": "0.0.1",
}
`;

exports[`upgrades library version 2`] = `
"{
  \\"name\\": \\"vcs-automation-scripts\\",
  \\"version\\": \\"0.0.1\\",
  \\"description\\": \\"VCS automation scripts\\",
  \\"repository\\": {
    \\"type\\": \\"git\\",
    \\"url\\": \\"https://github.com/Microsoft/TypeScript-Node-Starter\\"
  },
  \\"author\\": \\"Alexey Samoshkin\\",
  \\"license\\": \\"MIT\\",
  \\"scripts\\": {
    \\"build\\": \\"tsc\\",
    \\"buildw\\": \\"tsc -w\\",
    \\"lint\\": \\"tsc --noEmit && eslint\\",
    \\"start\\": \\"node -r ts-node/register src/main.ts\\",
    \\"test\\": \\"jest --forceExit --verbose\\"
  },
  \\"devDependencies\\": {
    \\"@types/jest\\": \\"^27.0.0\\",
    \\"@types/node\\": \\"^18.15.3\\",
    \\"@types/semver\\": \\"^7.5.5\\",
    \\"@types/shelljs\\": \\"^0.8.15\\",
    \\"@types/tmp\\": \\"^0.2.6\\",
    \\"@typescript-eslint/eslint-plugin\\": \\"4.14.2\\",
    \\"@typescript-eslint/parser\\": \\"4.14.2\\",
    \\"eslint\\": \\"7.19.0\\",
    \\"jest\\": \\"^27.0.6\\",
    \\"shelljs\\": \\"1.0.0\\",
    \\"ts-jest\\": \\"^27.0.3\\",
    \\"type-fest\\": \\"^4.8.1\\",
    \\"typescript\\": \\"4.9\\"
  },
  \\"engines\\": {
    \\"node\\": \\">=18.0.0\\"
  },
  \\"dependencies\\": {
    \\"bitbucket\\": \\"^2.11.0\\",
    \\"semver\\": \\"^7.5.4\\",
    \\"tmp\\": \\"^0.2.1\\",
    \\"ts-node\\": \\"^10.9.1\\"
  }
}"
`;
